test/bin.js

```diff
@@ -0,0 +1,203 @@
+      killed = true
+      sbot.kill('SIGINT')
+    })
+  }, 1200)
+})
+
+if (has_ipv6)
+test('run bin.js server with command line option --host and --port (IPv6)', function(t) {
+  var sbot = spawn(join(__dirname, '../bin.js'),  [
+    'server',
+    '--host=::1',
+    '--port=8199',
+    '--ws.port=8299',
+    '--path=/tmp/' + Date.now()
+  ], {
+    env: Object.assign({}, process.env, {ssb_appname: 'test'}),
+    stdio: 'inherit'
+  })
+
+  sbot.on('exit', function(code) {
+    t.notOk(code, 'sbot exit code should be zero')
+    t.end()
+  })
+
+  setTimeout( function() {
+    var socket = net.connect(8199, '::1')
+    socket.on('error', function(err) {
+      t.fail(err)
+      sbot.kill('SIGINT')
+    })
+    socket.on('connect', function(err) {
+      t.pass('connected successfully')
+      sbot.kill('SIGINT')
+    })
+  }, 1200)
+})
+
+test('run bin.js server with local config file (port, host) (IPv4)', function(t) {
+  var dir = '/tmp/' + Date.now()
+  mkdirp.sync(dir)
+  fs.writeFileSync(join(dir, '.testrc'), JSON.stringify({
+    host: '127.0.0.1',
+    port: 8399,
+    ws: {
+      port: 8499
+    }
+  }))
+  var sbot = spawn(join(__dirname, '../bin.js'),  [
+    'server',
+    `--path=${dir}`
+  ], {
+    env: Object.assign({}, process.env, {ssb_appname: 'test'}),
+    cwd: dir,
+    stdio: 'inherit'
+  })
+
+  sbot.on('exit', function(code) {
+    t.notOk(code, 'sbot exit code should be zero')
+    t.end()
+  })
+
+  setTimeout( function() {
+    var socket = net.connect(8399, '127.0.0.1')
+    socket.on('error', function(err) {
+      t.fail(err)
+      sbot.kill('SIGINT')
+    })
+    socket.on('connect', function(err) {
+      t.pass('connected successfully')
+      sbot.kill('SIGINT')
+    })
+  }, 1200)
+})
+
+if (has_ipv6)
+test('run bin.js server with local config file (port, host) (IPv6)', function(t) {
+  var dir = '/tmp/' + Date.now()
+  mkdirp.sync(dir)
+  fs.writeFileSync(join(dir, '.testrc'), JSON.stringify({
+    host: '::',
+    port: 8399,
+    ws: {
+      port: 8499
+    }
+  }))
+  var sbot = spawn(join(__dirname, '../bin.js'),  [
+    'server',
+    `--path=${dir}`
+  ], {
+    env: Object.assign({}, process.env, {ssb_appname: 'test'}),
+    cwd: dir,
+    stdio: 'inherit'
+  })
+
+  sbot.on('exit', function(code) {
+    t.notOk(code, 'sbot exit code should be zero')
+    t.end()
+  })
+
+  setTimeout( function() {
+    var socket = net.connect(8399, '::1')
+    socket.on('error', function(err) {
+      t.fail(err)
+      sbot.kill('SIGINT')
+    })
+    socket.on('connect', function(err) {
+      t.pass('connected successfully')
+      sbot.kill('SIGINT')
+    })
+  }, 1200)
+})
+
+test('sbot should have websockets and http server by default', function(t) {
+  var path = '/tmp/' + Date.now()
+  var killed = false
+  var sbot = spawn(join(__dirname, '../bin.js'),  [
+    'server',
+    '--host=127.0.0.1',
+    '--port=8199',
+    '--ws.port=0', // use random port - get them from getAddress and dial that
```

Comment:
We need at least one test that confirms that config.ws.port is actually used. Letting it generate its own address partially defeats that purpose. (we know it checked for zero, we don't know if it uses non-zero port values)

Reply:
Yes, agreed. I understand now. Using 0 to signal we want it to pick it's own won't tell us if it's broken and it just picks the default.

Reply:
In full transparency: I did this because sometimes the node process stuck around because the test script exited without killing it.

I'll try to use tape.onFinish to collect the PIDs and kill them collectively at the end.

Reply:
Ah, I see. Yes, that will do, as long as all the tests use different port numbers, so there's no ADDRINUSE when individual tests fail to kill sbot. It would be even better if we can have a finish handler that runs per test.